### PR TITLE
feat: add declare transactions

### LINKED ISSUE #32: Support Declare Transactions
### Overview

https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/definitions/transaction_type.py#L5

### Specification

May have to poke around Starknet.py or cairo-lang to fully understand.
It is for changing code or putting some code up or something like that...

### Dependencies

Include links to any open issues that must be resolved before this feature can be implemented.

### PR DESCRIPTION
**Pull request by ca11ab1e**
### What I did

(kind of started to) Add support for declare TX.

fixes #32 

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: ape_starknet/_utils.py [hunk 1]
```diff
@@ -117,3 +124,19 @@ def get_virtual_machine_error(err: Exception) -> Optional[VirtualMachineError]:
     # Fix escaping newline issue with error message.
     err_msg = err_msg.replace("\\n", "").strip()
     return ContractLogicError(revert_message=err_msg)
+
+
+def compute_contract_class_hash(path: Path) -> int:
```
Comment (ca11ab1e) [inline ape_starknet/_utils.py:129]: Raw implementation. A better approach would be to use what we know from the contract, maybe moving that function to a method at the good location.

File: README.md [hunk 2]
```diff
@@ -77,19 +80,80 @@ ape starknet accounts delete <ALIAS> --network starknet:testnet
...
+
+provider = networks.active_provider
+declaration = provider.declare(project.MyContract)
+print(declaration.class_hash)
+```
+
+Then, you can use the class hash in a deploy system call in a factory contract:
+
+```cairo
+@external
```
Comment (ca11ab1e) [inline README.md:101]: Maybe add the the proper import (or all required imports)?
```python
# at least
from starkware.starknet.common.syscalls import deploy
```

**Review by ca11ab1e — COMMENTED** [review]
:+1: Excellent!

File: ape_starknet/provider.py [hunk 3]
```diff
@@ -214,24 +228,51 @@ def send_call(self, txn: TransactionAPI) -> bytes:
...
-            yield self.network.ecosystem.create_transaction(**txn_dict)
+            yield self.starknet.create_transaction(**txn_dict)
 
     @handle_client_errors
     def send_transaction(self, txn: TransactionAPI, token: Optional[str] = None) -> ReceiptAPI:
+        txn_info = self._send_transaction(txn, token=token)
+        invoking = txn.type == TransactionType.INVOKE_FUNCTION
+
+        if "code" in txn_info and txn_info["code"] != StarkErrorCode.TRANSACTION_RECEIVED.name:
+            raise TransactionError(message="Transaction not receiver.")
```
Comment (unparalleled-js) [inline ape_starknet/provider.py:251]:
```suggestion
            raise TransactionError(message="Transaction not received.")
```

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**